### ISSUE TITLE
thanos receive: WAL files growing and no blocks were created for a month

### ISSUE THREAD

**Issue by hiepqdinh** [thanos-io/thanos#6995]
**Thanos, Prometheus and Golang version used**:
thanos v0.29.0
golang: v0.19
prometheus: v2.45 (I don't think it's relevant though)

**Object Storage Provider**:
AKS

**What happened**:
We have thanos-receive running and multiple pods using different tenant pushing metrics to it. There's `default-tenant` and `exporter-1` tenants. We just noticed that starting about a month ago (around November 16), there's no longer any blocks with `tenant_id="exporter-1"` in our bucket, querying metrics with this label still return data.

Looking closer to the thanos-receive pod, we see that under `/volumes/data/exporter-1` (the datadir is `/volumes/data`), the `wal` folder is now approaching 18Gi, it has more than a hundreds files, dated all the way back from Nov 16 to today.

We tried to restart the thanos receive but it starts up normally, albeit a bit slow (took 7 minutes because it needs to load 100+ WAL files), log is pasted below.

Shortly after the restart (7-8m of loading WAL files) it's ready and happily receive ingestion for both the good tenant (`default-tenant`) and the "bad" one (`exporter-1`), querying metrics for the bad-tenant also working fine. Which means those metrics are loaded, kept in, and served from memory.

Ssh into the pod and I see another `.tmp-for-creation` folder created under `/volumes/data/exporter-1/`

When I killed the pod, the last log entries for `tenant=exporter-1` just says:

```
level=info ts=... caller=multitsdb.go:231 component=receive component=multi-tsdb msg="flushing TSDB" tenant=exporter-1
```
but no block is actually created in the local PVC.

**What you expected to happen**:
blocks should be created and flushed from WAL data, and then uploaded to bucket storage, and then the WAL folder should be cleaned up (it shouldn't have 150 files)

**How to reproduce it (as minimally and precisely as possible)**:

**Full logs to relevant components**:
log after each restart is like this:
```
level=info ts=2023-12-19T00:01:06.640212614Z caller=multitsdb.go:415 component=receive component=multi-tsdb msg="a leftover lockfile found and removed" tenant=exporter-1
level=info ts=2023-12-19T00:01:06.641692635Z caller=multitsdb.go:498 component=receive component=multi-tsdb tenant=exporter-1 msg="opening TSDB"
level=info ts=2023-12-19T00:01:06.641906042Z caller=repair.go:56 component=receive component=multi-tsdb tenant=exporter-1 msg="Found healthy block" mint=1700157600005 maxt=1700164800000 ulid=[blockId1]
..
level=info ts=2023-12-19T00:01:06.642435867Z caller=db.go:851 component=receive component=multi-tsdb tenant=exporter-1 msg="Found and deleted tmp block dir" dir=/volumes/data/exporter-1/[BlockIdA].tmp-for-creation
level=info ts=2023-12-19T00:01:06.65118162Z caller=head.go:551 component=receive component=multi-tsdb tenant=exporter-1 msg="Replaying on-disk memory mappable chunks if any"
level=info ts=2023-12-19T00:01:08.574150722Z caller=head.go:595 component=receive component=multi-tsdb tenant=exporter-1 msg="On-disk memory mappable chunks replay completed" duration=1.922932153s
level=info ts=2023-12-19T00:01:08.574210723Z caller=head.go:601 component=receive component=multi-tsdb tenant=exporter-1 msg="Replaying WAL, this may take a while"
level=info ts=2023-12-19T00:01:08.586074354Z caller=head.go:637 component=receive component=multi-tsdb tenant=exporter-1 msg="WAL checkpoint loaded"
level=info ts=2023-12-19T00:01:09.246470012Z caller=head.go:672 component=receive component=multi-tsdb tenant=exporter-1 msg="WAL segment loaded" segment=2611 maxSegment=2759
...
level=info ts=2023-12-19T00:01:09.246470012Z caller=head.go:672 component=receive component=multi-tsdb tenant=exporter-1 msg="WAL segment loaded" segment=2759 maxSegment=2759
level=info ts=2023-12-19T00:09:36.622811124Z caller=head.go:709 component=receive component=multi-tsdb tenant=exporter-1 msg="WAL replay completed" checkpoint_replay_duration=11.898576ms wal_replay_duration=8m28.036689012s wbl_replay_duration=290ns total_replay_duration=8m29.971595811s
level=info ts=2023-12-19T00:09:36.628132029Z caller=multitsdb.go:528 component=receive component=multi-tsdb tenant=exporter-1 msg="TSDB is now ready"
```

**Anything else we need to know**:
For completeness, we have 3 thanos receive running with hashring, and the default hash algorithm, replicationFactor=1.

**Comment by GiedriusS** [comment]
That's an old version, there have been many, many fixes. Is it the same with the latest `main`?

**Comment by hiepqdinh** [comment]
So we found the issue. Apparently one of the push client was creating malformed metrics with duplicate labels pushing to our thanos-receive. For various legacy reason our thanos binary disable a subset of the checks in the receive path, among those are the dup label check. I ended up separating those WAL files (to savage whatever was still useful) to a set of “read only” receivers for awhile before completely removing them.
